File: src/wrover_status_light/src/DemoServer.py
#!/usr/bin/env python

import rospy
from wrover_status_light.srv import DoubleService
from wrover_status_light.srv import DoubleServiceRequest
from wrover_status_light.srv import DoubleServiceResponse


'''
The handler function of the service. It will be called whenever this server receives a new 
request message. Handler functions are responsible for carrying out the actual logics of the 
service and send a response back to the client. Handler functions in Python always take in a 
single argument of the request message type, and returns an instance of the response message type.
'''
def double_service_handler(request: DoubleServiceRequest):
    '''
    In this case, the handler function doubles the input integer and returns it. 
    '''
    return request.input * 2

    '''
    Another way to return a response is: 
    return DoubleServiceResponse(request.input * 2)
    However, in this case, returning a response can be simplified into what is used above 
    because it only contains a single field. Valid return types from handler functions are: 
    - None (which implies a failure)
    - Response type to a service
    - A tuple or list
    - A dictionary
    - The value of the field for response types with a single field
    '''

'''
Initializes the node
'''
def double_service_server():
    '''
    Creates a new node called "double_server". Without this the code cannot communicate with ROS, 
    and is usually declared at the beginning.
    '''
    rospy.init_node("double_server")

    '''
    Creates a new service called "double_service" that is defined by the "DoubleService" class, 
    and uses the function "double_service_handler" to handle any requests it receives.
    '''
    s = rospy.Service("double_service", DoubleService, double_service_handler)

    '''
    The rospy.spin() function and other versions of spin() keeps the code running by entering a 
    loop that listens for incoming requests.
    '''
    rospy.spin()

if __name__ == "__main__":
    double_service_server()

File: src/wrover_status_light/src/DemoServer.cpp
#include "ros/ros.h"
#include "ros/init.h"
#include <wrover_status_light/DoubleService.h>

/*
The handler function of the service. It will be called whenever this server receives a new 
request message. Handler functions are responsible for carrying out the actual logics of the 
service and send a response back to the client. Handler functions in C++ always takes in a 
reference to the request and a reference to the response as arguments.
*/
bool doubleServiceCallback(wrover_status_light::DoubleServiceRequest& req,
                        wrover_status_light::DoubleServiceResponse& res) {
    /*
    Here the handler functions doubles the input field for the requets and sets it as the output 
    field for the response.
    */
    res.output = req.input * 2;

    /*
    Since a reference to the response is already provided in the arguments, we don't need to 
    return it. Instead, handler functions in C++ return a boolean to convey if the operation is 
    successful or not. 
    */
    return res.output > 0;
}

int main(int argc, char** argv) {
    /*
    The ros::init() function sets up the node and needs to be called before using any other parts 
    of the ROS library. Here it takes in 3 arguments: argc, argv, and the name of this node. 
    */
    ros::init(argc, argv, "double_server");

    /*
    A NodeHandle is the main access point to communications with ROS in C++. Here it is used to 
    create a new server.
    */
    ros::NodeHandle nh;

    /*
    Creates a new server instance using the NodeHandle, with "double_service" as the server name 
    and "doubleServiceCallback" as the handler function. 
    */
    ros::ServiceServer service = nh.advertiseService("double_service", doubleServiceCallback);

    /*
    The ros::spin() function and other versions of spin() keeps the code running by entering a 
    loop that listens for incoming requests.
    */
    ros::spin();
    return 0;
}